0014-resume-swaps-after-restart.adoc

```diff
@@ -0,0 +1,110 @@
+That is, create if it does not exist or update if it does exist.
+
+How is this relevant for our usecase?
+
+Currently, we use POST requests to create the queries on btsieve.
+The difference is that POST requests, contrary to PUT requests do not make any guarantees about idempotence.
+In other words, a POST request will always create a new query whereas repeated PUT requests only create them on the first one and update afterwards.
+
+If the client wishes to use PUT requests to create new resources, it has to ensure that the used identifier is unique (otherwise it will issue an update rather than a create).
+In general, and contrary to POST requests, the client is in charge of creating identifiers.
+
+One thing to consider is that we may want to implement a verification step on btsieve that disallows updating a query with different content.
+Since this is application logic, it is still a valid usage of HTTP.
+
+== Recommendation
+
+We propose the following changes:
+
+1. Extend btsieve to allow the creation of queries through PUT (btsieve MUST validate that within repeated requests, the query is semantically equivalent)
+2. Change cnd to compute a unique identifier for each query.
+We propose to concat the secret hash with current query and hash it again:
+
+[source]
+----
+HASH_256(SECRET_HASH + "FUND")
+----
+
+With this change, we only have to slighly modify the creation of queries on the cnd side:
+
+.https://github.com/comit-network/comit-rs/blob/5895440ee7762d6f51001b7cf47d278fdb43b841/cnd/src/swap_protocols/rfc003/bitcoin/htlc_events.rs#L28-L50
+[source,diff]
+----
+diff --git a/cnd/src/swap_protocols/rfc003/bitcoin/htlc_events.rs b/cnd/src/swap_protocols/rfc003/bitcoin/htlc_events.rs
+index eb46c727..0aafc5c2 100644
+--- a/cnd/src/swap_protocols/rfc003/bitcoin/htlc_events.rs
++++ b/cnd/src/swap_protocols/rfc003/bitcoin/htlc_events.rs
+@@ -26,8 +26,9 @@ impl HtlcEvents<Bitcoin, BitcoinQuantity> for Arc<dyn QueryBitcoin + Send + Sync
+         htlc_params: HtlcParams<Bitcoin, BitcoinQuantity>,
+     ) -> Box<DeployedFuture<Bitcoin>> {
+         let query_bitcoin = Arc::clone(&self);
++        let identifier = htlc_params.deployed_identifier();
+         let deployed_future = self
+-            .create(BitcoinQuery::deploy_htlc(htlc_params.compute_address()))
++            .create(identifier, BitcoinQuery::deploy_htlc(htlc_params.compute_address()))
+             .and_then(move |query_id| query_bitcoin.transaction_first_result(&query_id))
+             .map_err(rfc003::Error::Btsieve)
+             .and_then(move |tx| {
+----
+
+With this change, we don't care if the query already exists on btsieve by the time we issue this request.
+Hence, we also don't need to store the URL or anything else in order to resume a swap.
+All we need to do is re-instantiate the state machine in the current state (retrieved from the database) and upon the first poll, this code will reconstruct the future with the existing query on btsieve.
```

Comment:
We should also think about if we have _enough_ data to recover easily and how to restore from it in a smart way? 

For example, think about this case: 
Bob's cnd restarts just before he redeems (redeem hasn't happened). 
What do we store in the DB? Is it only the data you mention above?

- Pair of ledgers
- Pair of assets
- Secret/SecretHash
- Expiry times
- Identities

Afaik, with only this information available, we will need to create (at least) 3 queries and wait for btsieve to find the results (meaning querying the blockchain again if they were deleted):
* query for alpha funded
* query for beta funded
* query for alpha redeemed

However, if we store the past events/results in the database, (e.g. alpha redeemed was seen prior the restart) such as 
* alpha funded transaction
* beta funded transaction
* alpha redeemed transaction 
Then we could recover faster/nicer.

What do you think?

Reply:
> (meaning querying the blockchain again if they were deleted):

The scope of the spike (in the ticket) states that we can assume that the queries are still available on btsieve.

> However, if we store the past events/results in the database, (e.g. alpha redeemed was seen prior the restart)

I would try to avoid this as much as possible and keep the whole process idempotent because I think it is easier to reason about.

> Then we could recover faster/nicer.

Haven't done any performance testing of the proposed solution but I'd be surprised if this takes very long. In the end, it is a couple of HTTP requests to btsieve which retrieves the existing results from the database. That should be pretty fast :)